### PR TITLE
Polar area rework

### LINKED ISSUE #49: Rework polar area chart.
I will rework the polar area chart next as it seems to be the most related to the pie and doughnut.

### PR DESCRIPTION
**Pull request by SeppPenner**
Fixes https://github.com/Joelius300/ChartJSBlazor/issues/49.

### CODE REVIEW

**Review by Joelius300 — CHANGES_REQUESTED** [review]
The changes so far are good.  

However you have to keep in mind that you have to implement [`LinearRadialAxis`](https://www.chartjs.org/docs/latest/axes/radial/linear.html) for this!  

You will keep the `Scale` property but make it of type `LinearRadialAxis`.  
The `LinearRadialAxis` has to contain [these configurations](https://www.chartjs.org/docs/latest/axes/radial/linear.html#configuration-options). On there you can find links to the what those configuration objects look like.  

Thoroughly look through [LineChart.Axes](https://github.com/Joelius300/ChartJSBlazor/tree/master/ChartJs.Blazor/ChartJS/LineChart/Axes). This should be your guideline.  
One importance difference to mention though is that there is a [`Scales`](https://github.com/Joelius300/ChartJSBlazor/blob/master/ChartJs.Blazor/ChartJS/LineChart/Scales.cs) class for `LineChart` which contains the `Axes`. In the polar chart, you only have one `Axis`, directly at the options level (inside `PolarAreaOptions`).

Also don't forget to move [LineChart.Axes.Axis](https://github.com/Joelius300/ChartJSBlazor/blob/master/ChartJs.Blazor/ChartJS/LineChart/Axes/Axis.cs) from `ChartJS.LineChart.Axes` to `ChartJS.Common` (file and namespace) and fix the namespace errors this might have as a result. This is now appropriate since it's not only for `LineChart` anymore.  

One last thing:  
Here it already uses the short version. That's because it wasn't done consistenly before. I think we should either make this into the long version or rename the rest. Since it's proposed to do this all at once in https://github.com/Joelius300/ChartJSBlazor/issues/28, I think we should make these into the long version before the great rename-massacre.

Otherwise good summaries. There are things to be adressed outside of this rework like https://github.com/Joelius300/ChartJSBlazor/issues/45 and https://github.com/Joelius300/ChartJSBlazor/issues/33 which I will look after later (doesn't concern this PR).

**Comment by Joelius300** [comment]
I missed some things in my review:  

You also have to move [LineChart.GridLine](https://github.com/Joelius300/ChartJSBlazor/blob/master/ChartJs.Blazor/ChartJS/LineChart/GridLine.cs) to the `Common` namespace since it's now used in multiple places (and probably shouldn't have been associated with `LineChart` before anyway).  

Also you have to implement a new class `Ticks`. This class has to be modeled after [these configurations](https://www.chartjs.org/docs/latest/axes/styling.html#tick-configuration). Watch out for included classes like `MinorTicks` and `MajorTicks` which will also have to be created.  

This `Ticks` class will be the new base class of [`LineChart.Axes.Ticks.CartesianTicks`](https://github.com/Joelius300/ChartJSBlazor/blob/master/ChartJs.Blazor/ChartJS/LineChart/Axes/Ticks/CartesianTicks.cs) but **do not** do this in this pull request. I will change this later on.

This is quite a lot to add. I will add some (hopefully helpful) reminders and informations below.

A few things:  

- You don't have to implement the callbacks
- You don't have to implement fontFamilies
- For fontStyle there is a StringEnum I think
- Don't forget the ColorUtil mention for colors
- For `LineHeight` I might implement a special `ObjectEnum`. For now put `int`
- Don't forget explicit default values for non-nullable types (eg. `FontSize` as `int` is 12)
- There are indexable options. Put arrays for now.

Since this is quite a lot, please portion the stuff you do in separate commits. It's hard to review when you change everything at once in one big commit. So put the tick stuff in a single one, put the namespace/file moving in a single one, etc.  

Thanks for helping me, I really appreciate it :)

**Comment by SeppPenner** [comment]
I will look trough this in detail later.

**Comment by SeppPenner** [comment]
Note to myself: Add https://github.com/Joelius300/ChartJSBlazor/issues/51#issuecomment-519868007 as well.

**Comment by SeppPenner** [comment]
* Moved LineChart.Axes.Axis from ChartJS.LineChart.Axes to ChartJS.Common
* Added radial axis classes
* Moved LineChart.GridLine to Common
* Not implemented the callbacks
* Not implemented the fontFamilies
* For fontStyle there is a StringEnum I think --> I haven't seen this one yet...
* For LineHeight I might implement a special ObjectEnum. For now put int: I have added it as double because the default is 1.2
* Don't forget explicit default values for non-nullable types (eg. FontSize as int is 12) --> Done.
* There are indexable options. Put arrays for now. --> Which do you mean?

> Here it already uses the short version. That's because it wasn't done consistenly before. I think we should either make this into the long version or rename the rest. Since it's proposed to do this all at once in #28, I think we should make these into the long version before the great rename-massacre.

We should maybe do #28 as next step already (even if some charts are not yet done)...

I hope, I didn't miss anything...

**Comment by SeppPenner** [comment]
#51 is fixed, too: https://github.com/Joelius300/ChartJSBlazor/pull/50/commits/15c0662ea72e4ab3e69217f19084c416992c1027.

**Review by Joelius300 — CHANGES_REQUESTED** [review]
## General

- You didn't have to do `FontFamily` (up until now I didn't include the property either). If we include it, it should be null so chart.js uses their default.
- `default(string)` is null so we don't necessarily need to explicitly set a default value. Chart.js will handle that for us. It's very important for value-types where `default` isn't null but `string` is a reference type. As long as we don't use nullable everywhere we should keep it this way (only do value-types).
- For `bool`s you always use "Gets or sets _a_ value indicating whether".. I think "_the_ value" would be better. I did not mention this in https://github.com/Joelius300/ChartJSBlazor/issues/22, my bad.
- We have actually always used `Ticks` instead of `Tick`. We should continue with that because it's a config for a _set of elements_ and the property is also always plural. Furthermore e.g. `PointLabel` seems to be a single element (from the name) but actually it's a config and applies to multiple point-labels. Other such cases include: Tick**s**, GridLine**s**, Scale**s** (only `LineChart`), PointLabel**s**, .. I'll add this to the conventions (https://github.com/Joelius300/ChartJSBlazor/issues/22) if I don't forget. You actually undid this in [20a851a](https://github.com/Joelius300/ChartJSBlazor/pull/50/commits/20a851ad007940cd35312e73c56703120016fc53).
- Do not add "class" after "sub-config of the <see cref="LinearRadialAxis"/>" if it's reasonable to say "a/an XX". E.g "a LinearRadialAxis" is fine (it's often reasonable to use say "a XX").
- Make your classes abstract. In many cases like `Ticks` and `Axis`, they should not be used directly, they're supposed to be inherited. You can usually add `abstract` to all the classes except the one you're actually implementing. So if you're working on `LinearRadialAxis` and you have to add `RadialAxis` and `Axis`, then every class except for the "top" one (`LinearRadialAxis`) should be abstract. There might be exceptions but I don't have any in mind currently.
- Very important one: **If there is a property of a value-type (int, double, bool,..) and in the chart.js docs it does not have a default value, it has to be nullable!** This is very important since in javascript there isn't a default value for `int`, it's just not defined if you don't define it. If we don't mark those properties as nullable, they're going to get a value and break the whole system. These introduce very breaking bugs. It's very important to avoid those. You best avoid those by being aware of them while carefully reading the chart.js docs.
- If you implement an option, it usually has something similar already. It's usually also very apparent when reading through the docs. If you implement some ticks, please go look if there are already ticks, how they are implemented, if there are base-classes, etc. This way you can not only speed up the process, you can also avoid annoying bugs and inconsistencies.
- Good work with the "As per documentation here ..". Those can be very helpful.

## TODO (after PR)

- Implement `FontStyle` as StringEnum (my bad, I thought that was already done)
- Indexable options (they really seem more and more necessary) - https://github.com/Joelius300/ChartJSBlazor/issues/45

## Tick

- You should rename `Tick` to `Ticks` since we use plural because it's the configuration for a set of elements
- It's actually the ~~tick~~_ticks_-subconfig of ~~a chart~~_an axis_ (in the summary). Every cartesian axis and the linear radial (so all radial) axes have a ticks subconfig.
- Add the base class (here: <see cref="Tick"/>) in parenthesis after "Omitted options are inherited from options above" for Major and Minor so it's more clear what that means.

## MajorTick

- You should rename `MajorTick` to `MajorTicks` (same applies to `MinorTick(s)`) since we use plural because it's the configuration for a set of elements. 
- Summary says minor instead of major
- Summary of Enabled has grammar error: whether ~~this~~_these_ options..

## GridLine

- You should rename `GridLine` to `GridLines` since we use plural because it's the configuration for a set of elements
- The summary should probably be "The grid lines sub-config of an axis". Same as for ticks, basically every axis has one even though they're not mentioned as common.

## AngleLine

- You should rename `AngleLine` to `AngleLines` since we use plural because it's the configuration for a set of elements
- This applies to the whole project but you've done it there. Do not add "class" after "sub-config of the <see cref="LinearRadialAxis"/>" if it's reasonable to say "a/an XX" so "a LinearRadialAxis" which is fine (it usually is).
- Summary of `Display`: This class doesn't represent a single `AngleLine` but rather a configuration for all angle-lines. You can also see this in the docs where it says "if true, angle lines are shown."

## LinearRadialAxis

- Summary of `Ticks`: Correct link but it still says "grid lines" instead of "ticks".

## LinearTick

- You should rename `LinearTick` to `LinearRadialTicks` since first of all we use plural because it's the configuration for a set of elements and secondly we already have `LinearCartesianTicks`.
- You have `Max`, `Min`, `StepSize`, `Precision`, `SuggestedMax` and `SuggestedMin` properties which are all value-types and don't have a chart.js default value. These **have to be nullable**! If you don't make those nullable you'll break the whole chart since they'll all default to 0 which will prevent you from displaying negative values and make trouble with the step-size and precision. 
- You have implemented a `Ticks` class but did not derive from it in you `LinearTicks` class. You should do that. Otherwise you need to explain to me why you didn't inherit that class.

## PointLabel

- You should rename `PointLabel` to `PointLabels` since we use plural because it's the configuration for a set of elements
- Summary of LineHeight has one "///" too many.

## PolarAreaOptions 

- Summary of Scale: That's not true, it's not "scales" here since it's only this one axis. It should say "The scale (axis) for this chart." or something else singular.


## Summing up

I know this is, once again, a lot of stuff. Maybe I'm too picky but I really think these changes should be done.  
I hope I didn't miss any errors (e.g. Documentation links).  

Thank you for contributing :) and don't take my criticism personally (I'm not experienced with these reviews).

**Comment by SeppPenner** [comment]
# General:
* Removed FontFamily
* Used `null` as default values for `string`
* Gets or sets a value indicating whether... --> Gets or sets the value indicating whether...
* Use plural: Well, the class was named `GridLine`, so assumed to use singular... --> Is changed now for Ticks, GridLines, Scales, PointLabels, LinearTicks, AngleLines
* Do not add "class" after "sub-config of the " if it's reasonable to say "a/an XX". E.g "a LinearRadialAxis" is fine (it's often reasonable to use say "a XX"). --> Done
* Made classes abstract

# Tick:
* Renamed to plural class name
* Fixed documentation: `The ticks-subconfig of an <see cref="Axis"/>.`
* Added reference to base class in documentation after `Omitted ...`

# MajorTicks and MinorTicks:
* Renamed to plural class name
* Fixed copied summary for MajorTicks (Wer kopiert, verliert :D)
* Fixed grammar error in MajorTicks: `Enabled`

# GridLine:
* Renamed to plural class name
* Updated summary

# AngleLine:
* Renamed to plural class name
* Do not add "class" after "sub-config of the..." --> Fixed in whole branch now.
* Made summary of AngleLines's display plural

# LinearRadialAxis:
* Fixed summary of `Ticks`

# LinearTick:
* Renamed `LinearRadialTicks`
* Made Max, Min, StepSize, Precision, SuggestedMax and SuggestedMin nullable.
* You have implemented a Ticks class but did not derive from it in you LinearTicks class. You should do that. --> I thought you wanted to integrate the inheritance later. `This Ticks class will be the new base class of LineChart.Axes.Ticks.CartesianTicks but do not do this in this pull request. I will change this later on.` --> I guess, I misunderstood that. It's fixed now.

# PointLabel:
* Renamed to plural class name
* Fixed documentation of LineHeight

# PolarAreaOptions:
* Fixed summary of Scale

# Summing up
I had some copy-paste and spelling errors. I think it's okay. Your suggestions were all valid ones (Most of the time this was just something I didn't do on purpose or overlooked or whatever). :)

Just tell me if there is anything else to change.

**Review by Joelius300 — CHANGES_REQUESTED** [review]
First of all thanks for all the work. So, I really don't want to be that person that never accepts a pull request but unfortunately I'm not pleased with some of the changes.  

Since there are a few general things I really have to get out of the way, let's do those first.  

# Changes for charts we're not currently reworking

I really have to empathize this again, Please **do not** change classes from other charts. It is fine to update common classes. If a common class-name changes, it's also fine to do this with some tool which will of course also affect other charts. This is good, otherwise the project wouldn't even compile anymore.  
However, you also changed default values and removed properties for other charts. Please do not do this. We really need to stay focused on one chart because all charts (except the ones we reworked) have severe problems and redundancies. If you author a change that wasn't necessary in a different chart, no matter how small, it shows change in that chart. We don't want that, we want to do one chart at a time. Also most charts that we have not yet reworked have a ton of issues (lots of redundancies, inconsistancies, properties with false locations, etc). We don't want to work on those until we really choose to rework those.   

I will list all the places you did this where I can't accept it (unless mentioned differently):  

- In https://github.com/Joelius300/ChartJSBlazor/pull/50/commits/0cd0183cca8e79549a5b6ebaced106dc7a85f832 you changed default values for multiple unrelated charts including scatter, line, bar (base) and radar
- I can accept https://github.com/Joelius300/ChartJSBlazor/pull/50/commits/8f80747d042f9c20a39127a21e3e1b70351231a0 because it's just a small thing in the summary without actual code-change.
- In https://github.com/Joelius300/ChartJSBlazor/pull/50/commits/6819b4f39fa2b673eebf76966ea8c875dfcd8fd7 you changed the name of `Scale` to `Scales` of radar-chart. We didn't even get to this chart yet, we'll probably remove that class anyway and there are dozens other things to do with radar-chart. Please do not invest any time in charts we're not currently reworking!

# Tick, Ticks, Tick configuration and their relationship

So the next big thing is about class names and summaries. It seems to me that you did not fully understand how tick, ticks and tick configuration should be used in naming. Other cases for this are angle line(s),  grid line(s) and point label(s).

I will use tick(s) as the example here but it counts for all the other cases as well.

## One single element

If we're referring to a relation between a tick and something else or a value for a tick (e.g. a tick-height if that was a thing) we use singular because a tick is a single element in the chart (yes there are many ticks but each ticks a single element and this property refers to _one_ tick). This usually comes to place in summaries _inside_ the configuration-class (ticks). A good example is [CartesianTicks.LabelOffset](https://github.com/Joelius300/ChartJSBlazor/blob/master/ChartJs.Blazor/ChartJS/LineChart/Axes/Ticks/CartesianTicks.cs#L23). The summary says "Distance in pixels to offset the label from the centre point of the tick.." It's not the centre point of multiple ticks, it's just about a single tick.  
There's also [CartesianTicks.AutoSkipPadding](https://github.com/Joelius300/ChartJSBlazor/blob/master/ChartJs.Blazor/ChartJS/LineChart/Axes/Ticks/CartesianTicks.cs#L17). Here it says "Padding between the ticks on the horizontal axis.." This is referring to the relationship between _two tick-elements_ so we of course use plural _ticks_ since we have multiple elements.

## Configuration and naming conventions

There are configurations which apply to _each element_ (here _each tick_). In the original, the decision was made to use the plural for those classes. That means the class which contains the _Tick configuration_ (which ofc applies to all _Ticks_) is calles _Ticks_ (and not Tick or TickConfiguration).  
Something to mention here as well is that (at least imo) it sounds/looks a lot better if you say `The .. sub-config of the tick-configuration.` instead of `.. of the Ticks.`. The best possible thing imo would be `The .. sub-config of the tick-configuration (see <see cref="Ticks"/>.`

Once again, I will update the guidelines with this as well (I really wish I had more time).  

# Abstract classes

Now onto the last one. I'm really sorry to ask this but do you know what `abstract` means in C#?  
I specifically said in my last review that you can usually make every class leading up to the one you're actually implementing `abstract`.  

Because it really seems like you need a good example I'll walk you through `MajorTicks` (please do not take any of this personally).  

Is `MajorTicks` supposed to be used directly or is it just an intermediate class which holds common properties of another class which will derive from `MajorTicks`?  
It's supposed to be used directly. You can see this in the [chart.js docs page](https://www.chartjs.org/docs/latest/axes/styling.html#major-tick-configuration) where there is no mention of a different sort of major ticks. Every sign points to major ticks being used directly. This means you cannot mark it `abstract` otherwise you will not be able to create an instance of it.  

Then next up. Does `MajorTicks` inherit any class which we could mark `abstract`?  
There are two parts to the answer for this question.  

First of all, there should be a class which is inherited by `MajorTicks`. Which one? `MinorTicks`. If you look at [the docs](https://www.chartjs.org/docs/latest/axes/styling.html#minor-tick-configuration) you can see the `MinorTicks` has the same properties as `MajorTicks` with the exception of `Enabled` which is only present in `MajorTicks`.  
So what now? Derive `MajorTicks` from `MinorTicks` and add a single property, the `Enabled` property, to `MajorTicks`.  

The `abstract` part here is correct but there is one good point why we shouldn't derive `MajorTicks` from `MinorTicks`. If we do, you'd be able to assign a `MajorTicks` instance to `Ticks.Minor` which is declared `MinorTicks`. This is of course possible because of the inheritance. We work around this by adding another _abstract_ class called `SubTicks` or something like that. This class would contain all the properties that are common between the two sub-ticks-classes (`MinorTicks` and `MajorTicks`). These just happen to be all the properties of `MinorTicks`.  
Then we'd derive both `MinorTicks` and `MajorTicks` from `SubTicks`. Now we can't assign `MajorTicks` to `MinorTicks` anymore so that problem is solved. The only thing that might bug someone is that `MinorTicks` is empty because all its properties are already in `SubTicks`.

Now to the second part. Now `MajorTicks` has a base class. As I said before, usually you can mark every class leading up to the one you're implementing `abstract`. This would mean you could mark `SubTicks` (the base-class) `abstract`. Is that right? Yes. Why? The `SubTicks` class is only used to reduce redundancies and is never supposed to be used directly, so you don't need to be able to instanciate it.  
If the base-class were `MinorTicks`, we could not mark the base-class (now `MinorTicks`) `abstract`. Why? In `Ticks` you're supposed to use `MinorTicks` directly so you'd need to be able to instanciate a `MinorTicks`-instace.  

I hope you understood what I mean with all of this (I'm not the best "explainer").  


Okay now, let's go through this commit by commit (thanks for separating them, it really helps reviewing). They are in historical order.

## FontFamily and String-default value

As mentioned above, I sadly can't accept that commit. You made changes in other charts that didn't need to be made. Please do not do that.

## a / the value whether..

Also mentioned above, this is fine because there are no code changes.

## Making names plural

This commit has a few issues, sadly too many for me to be able to accept.    
First of all you changed stuff in other charts, which you really cannot do.  
Secondly you changed "tick labels" to "ticks labels" (and tick(s) marks) which doesn't make any sense. Tick labels (and marks) was perfectly fine and makes sense since it's the label/mark of a single tick and this option applies to all (plural) _tick labels_. This was my main motivation for the point about Tick, Ticks and Tick configuration.  

Changes from `GridLine` to `GridLines` were good. Same for Major and Minor `Ticks` and `Ticks` itself. You missed `AngleLines` however (you fix this in a later commit, all good).

## Description change

There were two good ones:

- Adding `A linear readial axis.` for LinearRadialAxis. Small thing though, Please put the "As per documentation here.." in a `<para></para>` when you add actual descriptions to those.
- Adding full stop after sentence.

All the other ones were not good. The original descriptions were fine (not perfect but definitely better than what you changed them to). I saw that you undo these changes later on in other commits.

- Ticks
- AngleLine
- LinearTicks
- PointLabels
- GridLines

## Make abstract

Here is the `abstract` clusterfuck. This commit was the motivation for my "rant" above.  
Also as mentioned above, I would factor out the common properties in another class (`Subticks`) and derive `MinorTicks` and `MajorTicks` from that.  

The good changes in this commit were:

- Renaming `AngleLine` to `AngleLines` which was missed before
- Making `Ticks` `abstract`

All the others (adding `abstract` to the following classes) can't go through as they prohibit the user from customizing polar-chart almost entirely. All of these classes are supposed to be used directly and don't just hold common properties and thus can't be `abstract`. 

- MinorTicks
- MajorTicks
- AngleLines
- LinearTicks
- PointLabels

## More description changes

Same as the other commit. Don't worsen okay descriptions.

Good changes here:

- Change `this` to `these` (grammar error)
- Undo bad description for `Ticks`
- Adding reference to "above options" (Ticks). This is not something that would prevent me from accepting the PR but the "See" should be lowercase. Again, that's not a big issue.

## Undo bad description change from earlier 

Perfect. 

Done: 

- GridLines

## Undo more bad descriptions

Perfect.

Done: 

- AngleLines
- LinearTicks
- PointLabels

## Made summary of AngleLines' display plural

Perfect.

## Fix copy error in LinearRaidalAxis.Ticks

Perfect.

## Renamed LinearTicks to LinearRadialTicks

Perfect.

## Make properties nullable in LinearRadialTicks and fix inheritance

Very good.  
It's still `abstract` but the changes are all perfect.

## Remove unnecessary /// in PointLabels

Perfect.

## Change false plural for scale

Perfect.

## Rebuild for xml-docs

Very good. In the best case you'd do that before every commit but since this was one chunk of work, it's fine.



Alrighty.  

This is definitely a big one. I honestly don't know what the best way to fix the issues would be because the first few commits make trouble but after that it's all good.  

Again, I really need to mention that I don't mean to offend you or anything. It's hard for me to review these things because I'm not the best at explaining things and, as mentioned a few times now, I have a rather tight schedule and this is a freetime project.  
I really appreciate the time you take to help me but please take the things I say serious (a few things I've mentioned many times now) and go through all the stuff I said. This also includes looking at the current code for simmilar classes (look at the already reworked classes).  
Also with the `abstract` stuff, I think you should have invested a few minutes to think about what makes sense to mark as abstract and what doesn't.

**Comment by SeppPenner** [comment]
# Changes for charts we're not currently reworking

* I disagree on this one (Default values are now unset and they were wrong before either...), but I will change that:

> In 0cd0183 you changed default values for multiple unrelated charts including scatter, line, bar (base) and radar

* Well, documentation changes in other classes: I wanted it to be consistent and prevent that we forget this later...

> I can accept 8f80747 because it's just a small thing in the summary without actual code-change

* Undone: 6819b4f 

> In 6819b4f you changed the name of Scale to Scales of radar-chart. We didn't even get to this chart yet, we'll probably remove that class anyway and there are dozens other things to do with radar-chart. Please do not invest any time in charts we're not currently reworking!

# Tick, Ticks, Tick configuration and their relationship

## One single element
Should be done now. Re-check this, please.

## Configuration and naming conventions
Should be done now. Re-check this, please.

# Abstract classes
> Now onto the last one. I'm really sorry to ask this but do you know what abstract means in C#?
I specifically said in my last review that you can usually make every class leading up to the one you're actually implementing abstract.

I honestly had no idea what you wanted to do with `abstract` classes here.

> I hope you understood what I mean with all of this (I'm not the best "explainer").

I have adjusted this, please review this as well.

## Commits
# FontFamily and String-default value
Done before already (First commit or so).

# Making names plural
Fixed. (I didn't see ticks marks anywhere).

# Description change
* Done:

> Adding A linear readial axis. for LinearRadialAxis. Small thing though, Please put the "As per documentation here.." in a <para></para> when you add actual descriptions to those.

* Are they ok then?

> All the other ones were not good. The original descriptions were fine (not perfect but definitely better than what you changed them to). I saw that you undo these changes later on in other commits.

# Make abstract

Done before, I undid this. Ok, I made Ticks abstract again here.

# More description changes
Made `See` lower case in Ticks.

# The rest
Okay already, `LinearRadialTicks` isn't abstract anymore (Some commit before already).

# Rebuild for xml-docs
I cannot recompile the project again now because I updated my system to Blazor-preview8 already and it throws the known errors like "Update OnInitAsync to OnInitializedAsync" and stuff like that (I didn't want to do this in this pull request). So, if there are errors, you need to tell me or correct them on your own if you have still preview7 installed.
